### ISSUE TITLE
[Feature Request] Three arguments hypot

### ISSUE THREAD

**Issue by giordano** [JuliaLang/julia#15542]
It would be nice to have a three arguments `hypot` function to compute 3D distances.  GNU Scientific Library [has it since 2007](http://git.savannah.gnu.org/cgit/gsl.git/commit/?id=e1711c84f7ba5c2b22d023dcb7e10810233fff27) and it has been introduced in [C++17 standard](http://en.cppreference.com/w/cpp/numeric/math/hypot).

I can try and implement it using `hypot{T<:AbstractFloat}(x::T, y::T)` in `base/math.jl` as template.

**Comment by StefanKarpinski** [comment]
Seems reasonable. There's an argument to be made for the n-argument version as well, but obviously special-casing for a small number of dimensions makes sense. Interestingly, `hypot(x) == abs(x)`.

**Comment by giordano** [comment]
Before creating a pull request, here is what I wrote.

```julia
# 1-arg hypot
hypot(x::Real) = abs(promote(float(x))...)

# 3-arg hypot
hypot(x::Real, y::Real, z::Real) =
    hypot(promote(float(x), float(y), float(z))...)
function hypot{T<:AbstractFloat}(x::T, y::T, z::T)
    x = abs(x)
    y = abs(y)
    z = abs(z)
    m = maximum([x, y, z])
    if isinf(m)
        return m
    elseif m == 0
        o = one(m)
        return m*sqrt((x/o)*(x/o) + (y/o)*(y/o) + (z/o)*(z/o))
    else
        return m*sqrt((x/m)*(x/m) + (y/m)*(y/m) + (z/m)*(z/m))
    end
end

# n-arg hypot
hypot(w::Real, x::Real, y::Real, z::Real...) =
    hypot(promote(float(w), float(x), float(y), z...)...)
function hypot(w::AbstractFloat, x::AbstractFloat,
               y::AbstractFloat, z::AbstractFloat...)
    a = abs([w, x, y, z...])
    m = maximum(a)
    if isinf(m)
        return m
    elseif m == 0
        r = a/one(m)
    else
        r = a/m
    end
    return m*sqrt(r'*r)[1]
end
```

Comments welcome.  For example, I don't know how to splice `z` in `promote` for the n-arg case.

Here are some successful tests:
``` julia
@test_approx_eq hypot(pi) pi
@test_approx_eq hypot(3, -4, 12) 13
@test_approx_eq hypot(1, -2, 3, -6, 7, -1) 10
```

**Comment by giordano** [comment]
> I don't know how to split `z` in `promote` for the n-arg case.

I think this should do the trick
```julia
float([z...])...
```

**Comment by andreasnoack** [comment]
- In the three argument version you probably want to avoid allocating an array by using `max(x,y,z)`.
- Maybe only scale when there is a risk for overflow. Divisions are expensive.
- When scaling, use `abs2(x/m)` instead of `(x/m)*(x/m)` to be sure that only divide once.
- In the general case (`n>3`) you can call `vecnorm` on the tuple. That will take care of scaling correctly.
- Just go ahead and open a PR. It's easier to discuss the code there.

**Comment by giordano** [comment]
@andreasnoack Thank you so much for your suggestions!  Bear with me, I started learning Julia a couple of weeks ago and this is my first contribution to Julia core, I'm still a newbie `:-)`

> * Just go ahead and open a PR. It's easier to discuss the code there.

Just done, see pull request linked above.

**Comment by giordano** [comment]
Fixed by pull request #15649.
